### PR TITLE
feat: add dns cache

### PR DESCRIPTION
**Pull request by DarkGL**
## This relates to...

https://github.com/nodejs/undici/pull/2552

## Rationale

This is version of dns cache with resolved merge conflicts and tests migrated to `node:test`, most of the work is done by @antoinegomez

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by DarkGL** [comment]
I need help with fixing test on mac os (works on windows and linux)

Test:

`
test('http example', async (t) => {
  t = tspl(t, { plan: 1 })

  const cacheable = new DNSResolver({ resolver })

  const options = {
    hostname: 'example',
    port: 9999,
    lookup: cacheable.lookup
  }

  await t.rejects(makeRequest(options), {
    message: 'connect ECONNREFUSED 127.0.0.127:9999'
  })
})

const makeRequest = (options) =>
  new Promise((resolve, reject) => {
    http.get(options, resolve).once('error', reject)
  })
`

on mac os this test fails with:

`
✖ test/dns-resolver.js (30015.227ms)
  'test timed out after 30000ms'
`

on windows and linux this test resolves correctly

File: README.md [hunk 1]
```diff
@@ -390,6 +390,61 @@ Returns: `URL`
...
+new Agent({
+  dnsResolver: false
+})
+```
+
+Provide your custom DNS resolver, it must implement the method `lookup`:
+
+```js
+new Agent({
+  dnsResolver: new MyCustomDNSResolver()
```
Comment (ronag) [inline README.md:433]: Why this? I believe using the existing connect option should be sufficient?
```suggestion
  connect: { lookup: myCustomDNSResolverLookup }
```

**Review by ronag — CHANGES_REQUESTED** [review]
I don't think this is necessary. Already possible:

```js
import CacheableLookup from 'cacheable-lookup';

const cacheable = new CacheableLookup(opts)

const agent = new Agent({
  connect: { lookup: cacheable.lookup }
})

**Comment by DarkGL** [comment]
Mentioning reviewers from orginal pr @mcollina @ShogunPanda @metcoder95 @KhafraDev

**Comment by KhafraDev** [comment]
As ronag mentioned it's already possible to use cacheable-lookup

**Comment by KhafraDev** [comment]
The only reason I could see for adding it is if we wanted to enable it by default, but we're probably better off just documenting it

**Comment by DarkGL** [comment]
Should we add it here https://github.com/nodejs/undici/tree/main/docs/examples or https://github.com/nodejs/undici/tree/main/docs/docs/best-practices ?

**Comment by ronag** [comment]
Examples

**Pull request closed without merging**